packages/task/src/browser/task-configuration-manager.ts

```diff
@@ -0,0 +1,200 @@
+import URI from '@theia/core/lib/common/uri';
+import { Emitter, Event } from '@theia/core/lib/common/event';
+import { EditorManager, EditorWidget } from '@theia/editor/lib/browser';
+import { PreferenceService } from '@theia/core/lib/browser';
+import { QuickPickService } from '@theia/core/lib/common/quick-pick-service';
+import { WorkspaceService } from '@theia/workspace/lib/browser/workspace-service';
+import { TaskConfigurationModel } from './task-configuration-model';
+import { TaskCustomization, TaskConfiguration } from '../common/task-protocol';
+import { WorkspaceVariableContribution } from '@theia/workspace/lib/browser/workspace-variable-contribution';
+import { FileSystem, FileSystemError } from '@theia/filesystem/lib/common';
+import { FileChange, FileChangeType } from '@theia/filesystem/lib/common/filesystem-watcher-protocol';
+import { PreferenceConfigurations } from '@theia/core/lib/browser/preferences/preference-configurations';
+
+@injectable()
+export class TaskConfigurationManager {
+
+    @inject(WorkspaceService)
+    protected readonly workspaceService: WorkspaceService;
+
+    @inject(EditorManager)
+    protected readonly editorManager: EditorManager;
+
+    @inject(QuickPickService)
+    protected readonly quickPick: QuickPickService;
+
+    @inject(FileSystem)
+    protected readonly filesystem: FileSystem;
+
+    @inject(PreferenceService)
+    protected readonly preferences: PreferenceService;
+
+    @inject(PreferenceConfigurations)
+    protected readonly preferenceConfigurations: PreferenceConfigurations;
+
+    @inject(WorkspaceVariableContribution)
+    protected readonly workspaceVariables: WorkspaceVariableContribution;
+
+    protected readonly onDidChangeEmitter = new Emitter<void>();
+    readonly onDidChange: Event<void> = this.onDidChangeEmitter.event;
+
+    protected readonly onChangedTaskConfigEmitter = new Emitter<FileChange>();
+    readonly onChangedTaskConfig: Event<FileChange> = this.onChangedTaskConfigEmitter.event;
+
+    @postConstruct()
+    protected async init(): Promise<void> {
+        this.updateModels();
+        this.preferences.onPreferenceChanged(e => {
+            if (e.preferenceName === 'tasks') {
+                this.updateModels();
+            }
+        });
+    }
+
+    protected readonly models = new Map<string, TaskConfigurationModel>();
+    protected updateModels = debounce(async () => {
+        const roots = await this.workspaceService.roots;
+        const toDelete = new Set(this.models.keys());
+        for (const rootStat of roots) {
+            const key = rootStat.uri;
+            toDelete.delete(key);
+            if (!this.models.has(key)) {
+                const model = new TaskConfigurationModel(key, this.preferences);
+                model.onDidChange(() => this.onChangedTaskConfigEmitter.fire({ uri: key, type: FileChangeType.ADDED }));
+                model.onDispose(() => this.models.delete(key));
+                this.models.set(key, model);
```

Comment:
Should not we fire an added event here? The first event happens in the constructor, `onDidChange` is registered too late for it.